diff --git a/src/interface/swarm.cpp b/src/interface/swarm.cpp
index 39f1e8a28..2b6baf405 100644
--- a/src/interface/swarm.cpp
+++ b/src/interface/swarm.cpp
@@ -78,8 +78,8 @@ Swarm::Swarm(const std::string &label, const Metadata &metadata, const int nmax_
       neighbor_received_particles_("neighbor_received_particles_", NMAX_NEIGHBORS),
       cell_sorted_("cell_sorted_", nmax_pool_),
       buffer_sorted_("buffer_sorted_", nmax_pool_), mpiStatus(true) {
-  PARTHENON_REQUIRE_THROWS(typeid(Coordinates_t) == typeid(UniformCartesian),
-                           "SwarmDeviceContext only supports a uniform Cartesian mesh!");
+  //PARTHENON_REQUIRE_THROWS(typeid(Coordinates_t) == typeid(UniformCartesian),
+  //                         "SwarmDeviceContext only supports a uniform Cartesian mesh!");
 
   uid_ = get_uid_(label_);
 
