diff --git a/src/utils/reductions.hpp b/src/utils/reductions.hpp
index ecce5cef..439d550e 100644
--- a/src/utils/reductions.hpp
+++ b/src/utils/reductions.hpp
@@ -68,7 +68,7 @@ struct ReductionBase {
     // Store the communicator in a shared_ptr, so that
     // MPI_Comm_free is called, but only when the last
     // copy of this ReductionBase is destroyed.
-    pcomm = std::shared_ptr<MPI_Comm>(new MPI_Comm, MPI_Comm_disconnect);
+    pcomm = std::shared_ptr<MPI_Comm>(new MPI_Comm, MPI_Comm_free);
     PARTHENON_MPI_CHECK(MPI_Comm_dup(MPI_COMM_WORLD, pcomm.get()));
 #endif
   }

