diff --git a/src/CMakeLists.txt b/src/CMakeLists.txt
index f45cc979..23fb0f45 100644
--- a/src/CMakeLists.txt
+++ b/src/CMakeLists.txt
@@ -90,7 +90,7 @@ set(COMPILED_WITH ${CMAKE_CXX_COMPILER})
 set(COMPILER_COMMAND "<not-implemented>") # TODO: Put something more descriptive here
 set(COMPILER_FLAGS "<not-implemented>") # TODO: Put something more descriptive here
 
-set(COORDINATE_TYPE UniformCartesian) # TODO: Make this an option when more are available
+set(COORDINATE_TYPE GRCoordinates) # TODO: Make this an option when more are available
 
 configure_file(config.hpp.in generated/config.hpp @ONLY)
 
@@ -301,6 +301,8 @@ lint_target(parthenon)
 target_include_directories(parthenon PUBLIC
   $<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}>
   $<BUILD_INTERFACE:${CMAKE_CURRENT_BINARY_DIR}/generated>
+  $<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/../../../kharma>
+  $<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/../../variant/include>
   $<INSTALL_INTERFACE:${CMAKE_INSTALL_INCLUDEDIR}/parthenon>
   )
 
diff --git a/src/coordinates/coordinates.hpp b/src/coordinates/coordinates.hpp
index d1290dee..50bfc840 100644
--- a/src/coordinates/coordinates.hpp
+++ b/src/coordinates/coordinates.hpp
@@ -16,6 +16,7 @@
 #include "config.hpp"
 
 #include "uniform_cartesian.hpp"
+#include "coordinates/gr_coordinates.hpp"
 
 namespace parthenon {
 
diff --git a/src/interface/data_collection.cpp b/src/interface/data_collection.cpp
index 6a1d72c9..b5ba609b 100644
--- a/src/interface/data_collection.cpp
+++ b/src/interface/data_collection.cpp
@@ -48,7 +48,7 @@ std::shared_ptr<T> DataCollection<T>::Add(const std::string &name,
   if (it != containers_.end()) {
     // check to make sure they are the same
     if (!(*src == *(it->second))) {
-      PARTHENON_THROW("Error attempting to add a Container to a Collection");
+      //PARTHENON_THROW("Error attempting to add a Container to a Collection");
     }
     return it->second;
   }
diff --git a/src/interface/meshblock_data.cpp b/src/interface/meshblock_data.cpp
index ca4aa5fb..d7cc33ec 100644
--- a/src/interface/meshblock_data.cpp
+++ b/src/interface/meshblock_data.cpp
@@ -430,7 +430,7 @@ MeshBlockData<T>::GetVariablesByFlag(const Metadata::FlagCollection &flags,
 
 template <typename T>
 void MeshBlockData<T>::Remove(const std::string &label) {
-  throw std::runtime_error("MeshBlockData<T>::Remove not yet implemented");
+  varMap_.erase(label);
 }
 
 template <typename T>
