diff --git a/src/CMakeLists.txt b/src/CMakeLists.txt
index 81a8a1bd..65ba74f8 100644
--- a/src/CMakeLists.txt
+++ b/src/CMakeLists.txt
@@ -90,7 +90,7 @@ set(COMPILED_WITH ${CMAKE_CXX_COMPILER})
 set(COMPILER_COMMAND "<not-implemented>") # TODO: Put something more descriptive here
 set(COMPILER_FLAGS "<not-implemented>") # TODO: Put something more descriptive here
 
-set(COORDINATE_TYPE UniformCartesian) # TODO: Make this an option when more are available
+set(COORDINATE_TYPE GRCoordinates) # TODO: Make this an option when more are available
 
 configure_file(config.hpp.in generated/config.hpp @ONLY)
 
@@ -279,6 +279,8 @@ lint_target(parthenon)
 target_include_directories(parthenon PUBLIC
   $<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}>
   $<BUILD_INTERFACE:${CMAKE_CURRENT_BINARY_DIR}/generated>
+  $<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/../../../kharma>
+  $<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/../../variant/include>
   $<INSTALL_INTERFACE:${CMAKE_INSTALL_INCLUDEDIR}/parthenon>
   )
 
diff --git a/src/coordinates/coordinates.hpp b/src/coordinates/coordinates.hpp
index d1290dee..50bfc840 100644
--- a/src/coordinates/coordinates.hpp
+++ b/src/coordinates/coordinates.hpp
@@ -16,6 +16,7 @@
 #include "config.hpp"
 
 #include "uniform_cartesian.hpp"
+#include "coordinates/gr_coordinates.hpp"
 
 namespace parthenon {
 
